testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java

```diff
@@ -0,0 +1,268 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.centraldogma.testing.internal;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import org.slf4j.Marker;
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.classic.spi.IThrowableProxy;
+import ch.qos.logback.classic.spi.LoggerContextVO;
+import ch.qos.logback.core.Appender;
+import ch.qos.logback.core.UnsynchronizedAppenderBase;
+import ch.qos.logback.core.spi.AppenderAttachable;
+import ch.qos.logback.core.spi.AppenderAttachableImpl;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+import io.netty.util.internal.logging.Slf4JLoggerFactory;
+
+/**
+ * A <a href="https://logback.qos.ch/">Logback</a> {@link Appender} that adds {@code "(expected exception)"}
+ * to the log messages of expected exceptions and lower their logging level to {@link Level#DEBUG}
+ * if it's higher than that.
+ */
+public class ExpectedExceptionAppender extends UnsynchronizedAppenderBase<ILoggingEvent>
+        implements AppenderAttachable<ILoggingEvent> {
+
+    private static final String EXPECTED_EXCEPTION = "(expected exception)";
+
+    private static final ConcurrentMap<String, String> exceptionAndMessage = new ConcurrentHashMap<>();
+
+    static {
+        if (InternalLoggerFactory.getDefaultFactory() == null) {
+            // Can happen due to initialization order.
+            InternalLoggerFactory.setDefaultFactory(Slf4JLoggerFactory.INSTANCE);
+        }
+    }
+
+    private final AppenderAttachableImpl<ILoggingEvent> aai = new AppenderAttachableImpl<>();
+
+    /**
+     * Adds {@code "(expected exception)"} to the log message and lower its logging level to
+     * {@link Level#DEBUG} if the exception occurred from the {@code action} matches {@code throwable},
+     * and contains {@code message}.
+     */
+    public static void withExpectedException(Class<?> throwable, String message,
+                                             ThrowingRunnable action) throws Exception {
+        requireNonNull(throwable, "throwable");
+        withExpectedException(throwable.getName(), message, action);
+    }
+
+    /**
+     * Adds {@code "(expected exception)"} to the log message and lower its logging level to
+     * {@link Level#DEBUG} if the exception occurred from the {@code action} matches {@code throwable},
+     * and contains {@code message}.
+     */
+    public static void withExpectedException(String throwable, String message,
+                                             ThrowingRunnable action) throws Exception {
+        requireNonNull(throwable, "throwable");
+        requireNonNull(message, "message");
+        requireNonNull(action, "action");
+        exceptionAndMessage.put(throwable, message);
```

Comment:
Just curious. Is there any chance to put messages with the same `throwable` at the same time?

Reply:
I think it's OK~! :-)
I was intending to use this only in the scope of `run()` in `ThrowingRunnable`. If the code emits more than one expected exception, then it must be separated in different `run()`. This is used only in unit tests and it is executed sequentially, so it's ok.

Reply:
Ok! Thank you for explanation! :-)